File: app/src/main/java/com/pratamalabs/furqan/repository/FurqanDatabase.java
package com.pratamalabs.furqan.repository;

import android.content.Context;
import android.database.SQLException;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteException;
import android.database.sqlite.SQLiteOpenHelper;
import android.preference.PreferenceManager;
import android.util.Log;
import android.widget.Toast;

import com.pratamalabs.furqan.FurqanApp;
import com.pratamalabs.furqan.FurqanSettings;

import org.androidannotations.annotations.Bean;
import org.androidannotations.annotations.EBean;
import org.apache.commons.io.FileUtils;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

/**
 * Created by pratamalabs on 9/6/13.
 */
@EBean(scope = EBean.Scope.Singleton)
public class FurqanDatabase extends SQLiteOpenHelper {

    public final static String DATABASE_PATH = "/data/data/com.pratamalabs.furqan/databases/";
    private static final int DATABASE_VERSION = 4;
    private static String DATABASE_NAME = "furqan.sqlite";
    private static String DATABASE_VERSION_TAG = "dbversion.txt";
    private final Context dbContext;

    FurqanSettings settings = FurqanSettings.get();
    private SQLiteDatabase dataBase;


    public static FurqanDatabase get() {
        return FurqanDatabase_.getInstance_(FurqanApp.instance);
    }

    public FurqanDatabase(Context context) {
        super(context, DATABASE_NAME, null, DATABASE_VERSION);
        this.dbContext = context;
        // checking database and open it if exists
        if (checkDataBase()) {
            openDataBase();
        } else {
            try {
                this.getReadableDatabase();
                copyDataBase();
                this.close();
                openDataBase();

            } catch (IOException e) {
                throw new Error("Error copying database");
            }
            Toast.makeText(context, "Initial database is created", Toast.LENGTH_LONG).show();
        }
    }

    void upgrade1DbLink(SQLiteDatabase sqLiteDatabase) {
        String update1 = "UPDATE Sources SET DownloadLink = replace( DownloadLink, 'http://andikapratama.com/data/', 'https://raw.githubusercontent.com/andikapratama/furqan/master/app/QuranData/ExtraTranslations/' ), UpdateLink = replace( UpdateLink, 'http://andikapratama.com/data/', 'https://raw.githubusercontent.com/andikapratama/furqan/master/app/QuranData/ExtraTranslations/' ) WHERE Id IN (45,108,109,110,111)";
        sqLiteDatabase.execSQL(update1);
    }

    void addedIndonesianTransliteration(SQLiteDatabase sqLiteDatabase) {
        String update1 = "INSERT INTO Sources (Id,Type,Name,Author,DownloadLink,UpdateLink,LastModifiedDate,Language,ProviderName,Status,Ordering,TanzilId) " +
                "VALUES (111,?,?,?,?,?,?,?,?,?,111,?)";


        sqLiteDatabase.execSQL(update1, new Object[]{
                "Translation",
                "Transliterasi Indonesia",
                "Unknown",
                "https://raw.githubusercontent.com/andikapratama/furqan/master/app/QuranData/ExtraTranslations/transliterasiIndonesia.txt",
                "https://raw.githubusercontent.com/andikapratama/furqan/master/app/QuranData/ExtraTranslations/transliterasiIndonesia.txt",
                "2016-2-15",
                "Indonesian",
                "Unknown",
                "notdownloaded",
                "pratama.id.transliterasi"
        });
    }

    private void setDbVersion() throws IOException {
        String dbVersionPath = DATABASE_PATH + DATABASE_VERSION_TAG;
        FileUtils.write(new File(dbVersionPath), String.valueOf(DATABASE_VERSION));
    }

    private void copyDataBase() throws IOException {
        InputStream myInput = dbContext.getAssets().open(DATABASE_NAME);
        String outFileName = DATABASE_PATH + DATABASE_NAME;
        OutputStream myOutput = new FileOutputStream(outFileName);

        byte[] buffer = new byte[1024];
        int length;
        while ((length = myInput.read(buffer)) > 0) {
            myOutput.write(buffer, 0, length);
        }

        myOutput.flush();
        myOutput.close();
        myInput.close();

        setDbVersion();

        //set default folding
        PreferenceManager.getDefaultSharedPreferences(dbContext.getApplicationContext())
                .edit()
                .putBoolean("pratama.eng.jalalain", true)
                .putBoolean("pratama.eng.asbabalnuzul", true)
                .commit();
    }

    public void openDataBase() throws SQLException {
        String dbPath = DATABASE_PATH + DATABASE_NAME;
        dataBase = SQLiteDatabase.openDatabase(dbPath, null, SQLiteDatabase.OPEN_READWRITE);
        checkAndUpdateVersion(dataBase);
    }

    private void checkAndUpdateVersion(SQLiteDatabase sqLiteDatabase) {

        String dbVersionPath = DATABASE_PATH + DATABASE_VERSION_TAG;
        int dbLocalVersion;
        try {
            File file = new File(dbVersionPath);
            String version = FileUtils.readFileToString(file);
            dbLocalVersion = Integer.parseInt(version);
        } catch (Exception e) {
            e.printStackTrace();
            return;
        }

        if (dbLocalVersion < DATABASE_VERSION) {
            upgrade1DbLink(sqLiteDatabase);
            if (dbLocalVersion < 3) {
                addedIndonesianTransliteration(sqLiteDatabase);
            }
            try {
                setDbVersion();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    private boolean checkDataBase() {
        SQLiteDatabase checkDB = null;

        boolean exist = false;
        try {
            String dbPath = DATABASE_PATH + DATABASE_NAME;
            checkDB = SQLiteDatabase.openDatabase(dbPath, null,
                    SQLiteDatabase.OPEN_READONLY);
        } catch (SQLiteException e) {
            Log.v("db log", "database does't exist");
        }

        if (checkDB != null) {
            exist = true;
            checkDB.close();
        }
        return exist;
    }

    @Override
    public void onCreate(SQLiteDatabase sqLiteDatabase) {

    }

    @Override
    public void onUpgrade(SQLiteDatabase sqLiteDatabase, int i, int i2) {
    }
}

File: app/src/main/java/com/pratamalabs/furqan/repository/FurqanDatabase.kt
package com.pratamalabs.furqan.repository

import android.content.Context
import android.database.SQLException
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteException
import android.database.sqlite.SQLiteOpenHelper
import android.preference.PreferenceManager
import android.util.Log
import android.widget.Toast
import com.pratamalabs.furqan.FurqanApp
import com.pratamalabs.furqan.FurqanSettings
import org.apache.commons.io.FileUtils
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

/**
 * Created by pratamalabs on 9/6/13.
 */
class FurqanDatabase(private val dbContext: Context) : SQLiteOpenHelper(dbContext, DATABASE_NAME, null, DATABASE_VERSION) {

    internal var settings = FurqanSettings.get()
    private var dataBase: SQLiteDatabase? = null

    init {
        // checking database and open it if exists
        if (checkDataBase()) {
            openDataBase()
        } else {
            try {
                this.readableDatabase
                copyDataBase()
                this.close()
                openDataBase()

            } catch (e: IOException) {
                throw Error("Error copying database")
            }

            Toast.makeText(dbContext, "Initial database is created", Toast.LENGTH_LONG).show()
        }
    }

    internal fun upgrade1DbLink(sqLiteDatabase: SQLiteDatabase?) {
        val update1 = "UPDATE Sources SET DownloadLink = replace( DownloadLink, 'http://andikapratama.com/data/', 'https://raw.githubusercontent.com/andikapratama/furqan/master/app/QuranData/ExtraTranslations/' ), UpdateLink = replace( UpdateLink, 'http://andikapratama.com/data/', 'https://raw.githubusercontent.com/andikapratama/furqan/master/app/QuranData/ExtraTranslations/' ) WHERE Id IN (45,108,109,110,111)";
        sqLiteDatabase?.execSQL(update1);
    }

    internal fun addedIndonesianTransliteration(sqLiteDatabase: SQLiteDatabase?) {
        val update1 = "INSERT INTO Sources (Id,Type,Name,Author,DownloadLink,UpdateLink,LastModifiedDate,Language,ProviderName,Status,Ordering,TanzilId) " + "VALUES (111,?,?,?,?,?,?,?,?,?,111,?)"


        sqLiteDatabase!!.execSQL(update1, arrayOf<Any>("Translation", "Transliterasi Indonesia", "Unknown", "https://raw.githubusercontent.com/andikapratama/furqan/master/app/QuranData/ExtraTranslations/transliterasiIndonesia.txt", "https://raw.githubusercontent.com/andikapratama/furqan/master/app/QuranData/ExtraTranslations/transliterasiIndonesia.txt", "2016-2-15", "Indonesian", "Unknown", "notdownloaded", "pratama.id.transliterasi"))
    }

    @Throws(IOException::class)
    private fun setDbVersion() {
        val dbVersionPath = DATABASE_PATH + DATABASE_VERSION_TAG
        FileUtils.write(File(dbVersionPath), DATABASE_VERSION.toString())
    }

    @Throws(IOException::class)
    private fun copyDataBase() {
        val myInput = dbContext.assets.open(DATABASE_NAME)
        val outFileName = DATABASE_PATH + DATABASE_NAME
        val myOutput = FileOutputStream(outFileName)

        val buffer = ByteArray(1024)
        var length = myInput.read(buffer)
        while (length > 0) {
            length = myInput.read(buffer)
            myOutput.write(buffer, 0, length)
        }

        myOutput.flush()
        myOutput.close()
        myInput.close()

        setDbVersion()

        //set default folding
        PreferenceManager.getDefaultSharedPreferences(dbContext.applicationContext)
                .edit()
                .putBoolean("pratama.eng.jalalain", true)
                .putBoolean("pratama.eng.asbabalnuzul", true)
                .commit()
    }

    @Throws(SQLException::class)
    fun openDataBase() {
        val dbPath = DATABASE_PATH + DATABASE_NAME
        dataBase = SQLiteDatabase.openDatabase(dbPath, null, SQLiteDatabase.OPEN_READWRITE)
        checkAndUpdateVersion(dataBase)
    }

    private fun checkAndUpdateVersion(sqLiteDatabase: SQLiteDatabase?) {

        val dbVersionPath = DATABASE_PATH + DATABASE_VERSION_TAG
        val dbLocalVersion: Int
        try {
            val file = File(dbVersionPath)
            val version = FileUtils.readFileToString(file)
            dbLocalVersion = Integer.parseInt(version)
        } catch (e: Exception) {
            e.printStackTrace()
            return
        }

        if (dbLocalVersion < DATABASE_VERSION) {
            upgrade1DbLink(sqLiteDatabase)
            if (dbLocalVersion < 3) {
                addedIndonesianTransliteration(sqLiteDatabase)
            }
            try {
                setDbVersion()
            } catch (e: IOException) {
                e.printStackTrace()
            }

        }
    }

    private fun checkDataBase(): Boolean {
        var checkDB: SQLiteDatabase? = null

        var exist = false
        try {
            val dbPath = DATABASE_PATH + DATABASE_NAME
            checkDB = SQLiteDatabase.openDatabase(dbPath, null,
                    SQLiteDatabase.OPEN_READONLY)
        } catch (e: SQLiteException) {
            Log.v("db log", "database does't exist")
        }

        if (checkDB != null) {
            exist = true
            checkDB.close()
        }
        return exist
    }

    override fun onCreate(sqLiteDatabase: SQLiteDatabase) {

    }

    override fun onUpgrade(sqLiteDatabase: SQLiteDatabase, i: Int, i2: Int) {}

    companion object {

        val DATABASE_PATH = "/data/data/com.pratamalabs.furqan/databases/"
        private val DATABASE_VERSION = 4
        private val DATABASE_NAME = "furqan.sqlite"
        private val DATABASE_VERSION_TAG = "dbversion.txt"


        fun get(): FurqanDatabase {
            return FurqanDatabase(FurqanApp.instance)
        }
    }
}
